### PR TITLE
add password and email textbox types

### PR DESCRIPTION
**Pull request by pngwn**
# Description

Adds a new kwarg for `Textbox`, `type`. Defaults to `'text'` (current behaviour) but can optionally be set to `'password'` or `'email'`. `'email`' should autocomplete emails and could support patterns in the future. `'password'` blanks out the characters. Both `'email'` and `'password'` force the textbox to have a `maxlines` value of of 1 (because you can't have multiline password or email input types).

To test: add `type='password'` to any Textbox demo.

Closes half of #2029 but i don't actually want to close that issue for obvious reasons.

Please include: 
* relevant motivation
* a summary of the change 
* which issue is fixed. 
* any additional dependencies that are required for this change.

Closes: # (issue)

# Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have added a short summary of my change to the CHANGELOG.md
- [ ] My code follows the style guidelines of this project
- [ ] I have commented my code in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

# A note about the CHANGELOG

Hello 👋 and thank you for contributing to Gradio!

All pull requests must update the change log located in CHANGELOG.md, unless the pull request is labeled with the "no-changelog-update" label.

Please add a brief summary of the change to the Upcoming Release > Full Changelog section of the CHANGELOG.md file and include
a link to the PR (formatted in markdown) and a link to your github profile (if you like). For example, "* Added a cool new feature by `[@myusername](link-to-your-github-profile)` in `[PR 11111](https://github.com/gradio-app/gradio/pull/11111)`".

If you would like to elaborate on your change further, feel free to include a longer explanation in the other sections.
If you would like an image/gif/video showcasing your feature, it may be best to edit the CHANGELOG file using the 
GitHub web UI since that lets you upload files directly via drag-and-drop.

### CODE REVIEW

**Comment by pngwn** [comment]
why are you being like this

**Review by freddyaboulton — APPROVED** [review]
Looks good to me @pngwn ! 

I think it would be worthwhile to raise an exception on the backend during the textbox init if `type` is not "text", "email" or "password" otherwise the front-end will break. 

Might also be good to add a unit test about max_lines being set to 1 for password and email types. 

The auto complete for email doesn't seem to work for me (I hit tab and all it does it take me to another component). Not a dealbreaker for me but wanted to bring it up to your attention.

**Comment by pngwn** [comment]
@freddyaboulton The autocomplete is just the browser default, so in theory it should work however your browser's autocomplete/ autofill works but mine seems to be pretty inconsistent. Tabbing is still expected to change focus.

I'll do the other things.

**Comment by pngwn** [comment]
@freddyaboulton 
I added some tests for max_lines (with all types) and to check the error is correctly raised when a faulty type is passed.

lmk if they're okay, i have no idea what i'm doing.

File: test/test_components.py [hunk 1]
```diff
@@ -166,6 +164,18 @@ def test_override_template(self):
         assert component.get_config().get("value") == "abc"
         assert component.get_config().get("lines") == 4
 
+    def test_faulty_type(self):
```
Comment (freddyaboulton) [inline test/test_components.py:167]: Looks pretty good to me @pngwn !

Comment (freddyaboulton) [reply test/test_components.py:167]: I think you just need to modify gr.inputs.Textbox so that `type` is `text`

**Comment by pngwn** [comment]
ok we good

**Comment by pngwn** [comment]
Thanks for the review @freddyaboulton !

**Pull request merged**